contentcuration/contentcuration/utils/publish.py

```diff
@@ -819,62 +817,74 @@ def fill_published_fields(channel, version_notes):
     channel.save()
 
 
-def create_or_update_tsvectors(channel_id):
+def sync_contentnode_and_channel_tsvectors(channel_id):
     """
-    Create or update tsvectors for the channel and all its content nodes.
+    Creates, deletes and updates tsvectors of the channel and all its content nodes
+    to reflect the current state of channel's main tree.
```

Comment:
As you are marking all `ContentNodeFullTextSearch` in a channel with the `channel_id` you could just delete all existing ones here as a first and avoid a lot of checking here.

```
ContentNodeFullTextSearch.objects.filter(channel_id=channel_id).delete()
```

Reply:
I spend a good amount of time thinking about the approaches before moving ahead with the current one. Mainly there were two ways:

- delete all, insert all 
- delete, insert, update (current implementation)

In `delete all, insert all` there's a performance downside. For e.g. if the user has 10,000 nodes and he has added 10 more nodes & updated 100 nodes and then re-publishes, it'll delete & re-insert the already up to date 9,900 tsvectors. Keeping in mind the updates of GIN indexes and heavy cost of delete and insert, I decided to move ahead with the current implementation. 

Sir if we are ready to accept the performance downside of the mentioned approach then we can definitely change our current implementation.

Reply:
Yeah, I think with the correlated subquery below for updates, we can do one delete operation to remove no longer relevant nodes, and one operation to create the newly needed ones, so we can ignore this for now.